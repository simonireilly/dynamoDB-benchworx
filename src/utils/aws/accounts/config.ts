// Get Credentials from the local INI files and organize into types for sign in
// which are safe for persistance and use inside the DOM

import { PreloaderResponse } from "@src/preload";
import {
  loadSharedConfigFiles,
  Profile,
  SharedConfigFiles,
} from "@aws-sdk/shared-ini-file-loader";

export type SafeProfile = {
  profile: string;
  mfa: boolean;
  assumeRole: boolean;
  region: string;
  endpoint?: string;
};

export const listAwsConfig = async (): Promise<
  PreloaderResponse<SafeProfile[]>
> => {
  const config = await loadSharedConfigFiles();

  if (config) {
    const data = Object.entries(mergeCredentialsAndConfig(config)).map(
      safeConfigConstructor
    );

    console.info(data);

    return {
      type: "success",
      data,
      message: "Fetched local roles from INI files in ~/.aws",
      details: `Configurations found: ${data.length}`,
    };
  } else {
    return {
      type: "error",
      data: null,
      message: "Failed to fetch local roles from INI files in ~/.aws/",
      details: "",
    };
  }
};

const mergeCredentialsAndConfig = (config: SharedConfigFiles) => {
  const profiles: { [key: string]: Profile } = {};

  const credentialProfiles = Object.entries(config.credentialsFile).reduce(
    (acc, [key, values]) => {
      acc[key] = {
        ...values,
        ...config.configFile[key],
      };

      return acc;
    },
    profiles
  );

  const completeProfiles = Object.entries(config.configFile).reduce(
    (acc, [key, values]) => {
      acc[key] = {
        ...acc[key],
        ...values,
      };

      return acc;
    },
    credentialProfiles
  );

  return completeProfiles;
};

const safeConfigConstructor = (entry: [string, Profile]): SafeProfile => {
  const [profile, data = {}] = entry;
  const { aws_access_key_id, aws_secret_access_key, ...safeData } = data;

  const currentProfile: SafeProfile = {
    profile,
    mfa: Boolean(safeData.mfa_serial),
    assumeRole: Boolean(safeData.role_arn),
    region: safeData.region,
  };

  if (safeData.endpoint) currentProfile.endpoint = safeData.endpoint;

  return currentProfile;
};
