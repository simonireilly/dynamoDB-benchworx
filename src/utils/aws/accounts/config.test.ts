import { listAwsConfig } from "./config";

import { loadSharedConfigFiles } from "@aws-sdk/shared-ini-file-loader";
import { mocked } from "ts-jest/utils";

jest.mock("@aws-sdk/shared-ini-file-loader");
jest.spyOn(global.console, "info");

const fixture = {
  configFile: {
    simpleProfile: {
      region: "us-east-1",
    },
    default: {
      region: "eu-west-2",
      output: "json",
    },
    mfaRoleAssume: {
      mfa_serial: "arn:aws:iam::1111111111111:mfa/user.name",
      role_arn: "arn:aws:iam::2222222222:role/userRoleTOAssume",
      source_profile: "default",
      region: "eu-west-1",
      output: "json",
    },
    local: {
      region: "local",
      endpoint: "http://localhost:1234",
    },
    sso_admin: {
      sso_start_url: "https://sso.awsapps.com/start",
      sso_region: "eu-west-1",
      sso_account_id: "1123456789",
      sso_role_name: "AWSAdministratorAccess",
      region: "us-east-1",
      output: "json",
    },
  },
  credentialsFile: {
    simpleProfile: {
      aws_access_key_id: "access-1",
      aws_secret_access_key: "key-2",
    },
    cgu: {
      aws_access_key_id: "access-2",
      aws_secret_access_key: "key-2",
    },
    default: {
      aws_access_key_id: "access-3",
      aws_secret_access_key: "key-3",
    },
  },
};

const loadConfigMock = mocked(loadSharedConfigFiles);

describe("listAwsConfig", () => {
  it("loads the local config", async () => {
    loadConfigMock.mockResolvedValueOnce(fixture);
    const results = await listAwsConfig();

    expect(results.data).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          profile: "default",
          mfa: false,
          assumeRole: false,
          region: "eu-west-2",
        }),
        expect.objectContaining({
          profile: "mfaRoleAssume",
          mfa: true,
          assumeRole: true,
          region: "eu-west-1",
        }),
        expect.objectContaining({
          profile: "local",
          mfa: false,
          assumeRole: false,
          region: "local",
          endpoint: "http://localhost:1234",
        }),
      ])
    );
    expect(results.data.length).toEqual(6);
    expect(results.type).toEqual("success");
    expect(results.data).toMatchSnapshot();
  });

  it("returns an error response when there is no config", async () => {
    loadConfigMock.mockResolvedValueOnce(null);

    const results = await listAwsConfig();

    expect(results.type).toEqual("error");
    expect(results.message).toEqual(
      "Failed to fetch local roles from INI files in ~/.aws/"
    );
  });
});
