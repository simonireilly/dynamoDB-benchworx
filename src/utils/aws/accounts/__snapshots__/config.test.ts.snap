// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`listAwsConfig loads the local config 1`] = `
Array [
  Object {
    "assumeRole": false,
    "mfa": false,
    "profile": "simpleProfile",
    "region": "us-east-1",
  },
  Object {
    "assumeRole": false,
    "mfa": false,
    "profile": "cgu",
    "region": undefined,
  },
  Object {
    "assumeRole": false,
    "mfa": false,
    "profile": "default",
    "region": "eu-west-2",
  },
  Object {
    "assumeRole": true,
    "mfa": true,
    "profile": "mfaRoleAssume",
    "region": "eu-west-1",
  },
]
`;
